satpy/utils.py

```diff
@@ -631,6 +633,112 @@ def _get_pytroll_chunk_size():
         return None
 
 
+def chunks_by_resolution(
```

Comment:
Ok, I did quite some thinking about this, and I believe we can make this simpler and more flexible.
Basically the idea is to follow the signature of dask's `normalize_chunks` more closely. So that for example the signature of this function could be:
```python
def chunks_by_resolution(chunks, shape, low_res_multipliers, dtype, previous_chunks, high_res_shape):
```
where `chunks` would be eg `(-1, "auto")` for one polar data (-1 meaning full width chunk), `shape` is the shape of the current array, `low_res_multipliers` would be eg `(4, 4)` for 1km data, `previous_chunks` would be optional and related to on disk storage, and `high_res_shape` would be optional (good for the case where hi res data shape is not the low res data shape times the multiplier)
I believe this is all data that the reader has when calling this utility function, and using this signature, we allow having n-dimensional arrays without having to guess anything.
So with this, we can just pass "auto" and -1 from the chunks spec straight to normalize chunks. And I guess that would be the most used cases for the x and y dimensions right?
If we do that, maybe this function should be renamed to eg `normalize_low_res_chunks`?

Reply:
I'm OK with this I think. I'll have to see how it looks when I implement it. The main idea of allowing parameters for each dimension is worth it enough. I'm not sure I'm sold on the `high_res_shape` needing to be a parameter as I'd think that rounding any fractional results would produce the same thing.

Reply:
Ok. Major refactor done. Let me know what you think. I didn't add high res shape yet as I'm not sure I need it. Willing to be told otherwise.

Reply:
Oh I should have mentioned, I didn't pass previous chunks to dask because it gave me different results in my unit tests. I'll check on that in a bit and see if that's expected.

Reply:
Ok, added previous chunks now. This actually fixed something I hadn't noticed which was that chunks were being returned that weren't actually aligned to the disk chunking (previous chunks). At least not at the coarser resolutions.

I'm not sure I 100% like this logic, but I think it is good enough:

```python
low_res_chunks.append(int(max(hr_chunks[0] / lr_mult, prev_chunks / lr_mult)))
```

My main concern is that `prev_chunks / lr_mult` could produced a smaller chunk size that the on-disk chunking since `prev_chunks` could be coming from on-disk and it would be for this particular resolution. So if we took a on-disk chunk of 100 and got 100 / 4, then we'd be getting 4 dask chunks per on-disk chunk which has been shown to hurt performance pretty bad (ex. AGRI files). It could just be set to `prev_chunks` as a lower-limit, but then we could end up negating the resolution stuff...actually yeah I'm going to change this.